### PR TITLE
feat(events): implement numeric code check-in

### LINKED ISSUE #237: feat(events): participation module MVP
## Problem Statement

A He4rt Developers organiza meetups presenciais, workshops e conferências (incluindo um meetup em breve e uma conferência de 2 dias no final do ano), mas não possui um sistema próprio para gerenciar o ciclo de participação dos membros. Hoje não há como:
- Controlar inscrições com capacidade limitada e lista de espera
- Verificar presença de quem realmente compareceu (check-in)
- Recompensar participação com XP no sistema de gamificação existente
- Aprovar candidatos a eventos com vagas limitadas (workshops)

## Solution

Construir o módulo de participação em eventos (`app-modules/events/`) com lifecycle completo: enrollment → check-in → attendance → XP reward. O sistema suporta 3 métodos de entrada (RSVP, RSVP+check-in, Application com formulário dinâmico), 3 métodos de check-in (manual, código numérico, QR code), waitlist com promoção FIFO, e integração com o módulo de gamificação via domain events.

## User Stories

1. As a participant, I want to RSVP to a meetup with one click, so that I can confirm my presence quickly
2. As a participant, I want to join a waitlist when an event is full, so that I can be promoted if someone cancels
3. As a participant, I want to be automatically promoted from the waitlist when a slot opens, so that I don't miss the opportunity
4. As a participant, I want to cancel my enrollment before the deadline, so that my slot is freed for others
5. As a participant, I want to check in at a meetup by typing a numeric code, so that my presence is verified
6. As a participant, I want to check in at a conference by showing my QR code badge, so that attendance is tracked per day
7. As a participant, I want to receive XP when I confirm, check in, and complete attendance, so that my participation is rewarded in the gamification system
8. As a participant, I want to fill an application form to apply for a workshop, so that the organizer can evaluate my candidacy
9. As a participant, I want to be notified when my application is approved or rejected, so that I know my status
10. As a participant, I want to check in on each day of a multi-day conference, so that my attendance per day is tracked
11. As a participant, I want to use the same QR code across all event days, so that I only need one badge
12. As a participant, I want to check in via Discord/Twitch bot by typing a command with the code, so that I can confirm presence in online events
13. As a participant, I want to see my enrollment status for each event, so that I know where I stand
14. As an organizer, I want to create events with configurable enrollment policies, so that each event has appropriate participation rules
15. As an organizer, I want to set capacity limits with optional waitlist, so that I can control attendance
16. As an organizer, I want to choose the enrollment method (RSVP, RSVP+check-in, Application), so that the entry flow matches the event type
17. As an organizer, I want to choose the check-in method (manual, numeric code, QR code), so that presence verification fits the event format
18. As an organizer, I want to define a dynamic application form (JSONB schema), so that I can collect relevant information per event
19. As an organizer, I want to approve or reject applications with a reason, so that participants understand the decision
20. As an organizer, I want to generate time-limited numeric codes bound to a specific date, so that check-in is controlled per day
21. As an organizer, I want to manually check in participants, so that I can handle edge cases (late arrivals, code issues)
22. As an organizer, I want to override a participant's status (e.g., no_show → attended), so that I can correct errors
23. As an organizer, I want to set a cancellation deadline in hours before the event, so that I have predictability on attendance
24. As an organizer, I want to configure XP rewards per event (RSVP, check-in, attendance), so that different events reward differently
25. As an organizer, I want to set attendance_requirement (all_days, any_day, minimum_days), so that multi-day events have clear completion criteria
26. As an organizer, I want to see all enrollments with their current status, so that I can manage the event
27. As an organizer, I want a full audit trail of every status transition, so that I can investigate disputes
28. As the system, I want to automatically mark checked_in → attended after the event ends, so that XP is awarded without manual intervention
29. As the system, I want to automatically mark confirmed → no_show after the event ends, so that non-attendees are identified
30. As the system, I want enrollment capacity checks to be atomic (lockForUpdate), so that overbooking is impossible under concurrency

### Architecture

- **Module**: `app-modules/events/` following existing modular architecture (`internachi/modular`)
- **Namespace**: `He4rt\Events\*`
- **Panels**: Admin panel = organizer, App panel = participant
- **Table prefix**: All tables prefixed with `events_` for clear identification in database

### State Machine

- Implemented as a PHP backed enum (`EnrollmentStatusEnum`) with `canTransitionTo()` method
- **NOT** using `spatie/laravel-model-states` — transitions validated in Action classes explicitly
- 8 states: `pending`, `confirmed`, `waitlisted`, `checked_in`, `attended`, `cancelled`, `rejected`, `no_show`
- Terminal states: `attended` (success), `cancelled`, `rejected`, `no_show`
- See ADR-0001

### Schema (7 tables)

**`events`** — Core event record:
- `id`, `tenant_id`, `event_type` (meetup/workshop/conference), `active`, `slug`, `title`, `description`, `location`, `starts_at`, `ends_at`, `timestamps`
- No denormalized counters — derived by query

**`events_enrollment_policies`** — 1:1 configuration per event:
- `event_id` (unique), `enrollment_method`, `check_in_method`, `capacity` (nullable = unlimited), `has_waitlist`, `attendance_requirement` (all_days/any_day/minimum_days), `cancellation_deadline_hours`, `xp_reward_rsvp`, `xp_reward_checkin`, `xp_reward_attendance`, `application_form_schema` (JSONB), `timestamps`

**`events_enrollments`** — Hub central (1 per user+event):
- `id`, `event_id`, `user_id`, `status` (enum), `waitlist_position` (nullable), `application_data` (JSONB), `rejection_reason`, `enrolled_at`, `confirmed_at`, `checked_in_at`, `attended_at`, `cancelled_at`, `is_public`, `timestamps`
- UNIQUE constraint on `(event_id, user_id)`

**`events_enrollment_transitions`** — Append-only audit trail:
- `id`, `enrollment_id`, `from_status`, `to_status`, `triggered_by` (system/user/admin), `actor_user_id` (nullable), `reason` (nullable), `created_at`
- Written explicitly by Actions, NOT by database triggers (ADR-0003)

**`events_check_ins`** — 1:N per enrollment (one per day):
- `id`, `enrollment_id`, `method` (manual/numeric_code/qr_code), `payload` (JSONB), `event_date`, `checked_in_at`, `timestamps`
- Multi-day events: one check-in record per day attended (ADR-0004)
- Event days derived from `events.starts_at`/`ends_at` date range — no separate days table

**`events_check_in_codes`** — Organizer-generated numeric codes:
- `id`, `event_id`, `code`, `event_date`, `valid_from`, `valid_until`, `max_uses` (nullable), `uses_count`, `timestamps`
- Bound to specific date — one code per day

**`events_qr_tokens`** — One per enrollment, reusable across days:
- `id`, `enrollment_id`, `token` (unique), `expires_at` (nullable), `timestamps`

### XP Integration

- Events module publishes domain events: `EnrollmentConfirmed`, `ParticipantCheckedIn`, `ParticipantAttended`
- Gamification module listens and calls `IncrementExperience` on Character
- XP amounts come from `events_enrollment_policies` fields, passed as payload in domain events
- No XP ledger in Events module — auditability via `events_enrollment_transitions`
- Idempotency responsibility on Gamification listener
- See ADR-0002

### Bot Integration

- Discord/Twitch bot dispatches a domain event (e.g., `CheckInRequested`) with user identifier and code
- Events module has a listener that validates the code and processes the check-in
- Bot is pure transport — all business logic in Events module
- Bot must listen for response event (`CheckInProcessed`) to reply to user in chat
- Consistent with existing Bot Discord → Moderation pattern in CONTEXT-MAP
- See ADR-0005

### Enrollment Flow — Concurrency

- Capacity check uses `DB::transaction` with `lockForUpdate()` on the enrollment policy
- Count confirmed/checked_in/attended enrollments → compare against capacity → insert or waitlist
- Waitlist promotion on cancellation: FIFO (ordered by `enrolled_at`), atomic update + notification dispatch

### Scheduled Job

- Runs after `event.ends_at` passes
- Batch transitions: `checked_in → attended` (if attendance_requirement met), `confirmed → no_show`
- Dispatches XP domain events for newly `attended` enrollments
- Single job handles both transitions per event

### Application Form

- Dynamic form schema stored as JSONB in `events_enrollment_policies.application_form_schema`
- Filament form builder for organizer to define questions
- Participant responses stored in `events_enrollments.application_data` as JSONB
- Organizer reviews in Admin panel with approve/reject actions

### Event Types

```php
enum EventTypeEnum: string {
    case Meetup = 'meetup';
    case Workshop = 'workshop';
    case Conference = 'conference';
}
```

## Testing Decisions

- **Good tests**: test external behavior through Actions and domain events, not internal implementation. Assert state transitions, database state, and dispatched events.
- **Test location**: `app-modules/events/tests/Feature/` and `app-modules/events/tests/Unit/`
- **Prior art**: existing tests in `app-modules/moderation/tests/` and `app-modules/bot-discord/tests/`
- **What to test**:
  - Each Action (EnrollUser, CancelEnrollment, CheckIn, ApproveApplication, RejectApplication, PromoteFromWaitlist) — happy path + edge cases
  - State machine transitions — valid and invalid
  - Concurrency: capacity enforcement under race conditions
  - Waitlist promotion FIFO ordering
  - Multi-day check-in with attendance_requirement validation
  - Scheduled job (attended + no_show marking)
  - Domain event dispatch (assert events are fired, don't test Gamification's listener here)
  - Cancellation deadline enforcement
  - Numeric code validation (expiry, max_uses, wrong date)
  - QR token validation (expiry, reuse across days)
- **Factories**: create factories for EventModel, EventEnrollmentPolicy, EventEnrollment, EventCheckIn, EventCheckInCode, EventQrToken
- **Use Pest** with `php artisan make:test --pest`
- **ActingAs**: all Filament panel tests require `$this->actingAs(User::factory()->create())`

## Out of Scope

- Networking between participants (profiles, connections, discovery)
- Referral / invite links with tracking
- Magic link check-in method
- Geolocation check-in method
- Sponsors association with events
- Timeline / activity feed display
- Call for Papers / Submissions (CFP)
- Agenda / schedule display
- Paid events / payment integration
- No-show penalties affecting future eligibility (data tracked, consequences deferred)
- Event days as explicit table (derived from date range)
- Denormalized counters on events table

## Further Notes

- The events module was previously gutted (all tables dropped in migration `2026_05_16_195205`). This PRD rebuilds the module from scratch with the new participation-focused architecture.
- The previous module had: events, events_talks, events_attendees, sponsors, events_sponsors, events_agenda, event_submission_speakers. The new module replaces this with a fundamentally different schema focused on enrollment lifecycle.
- Existing ADRs document key architectural decisions: see `app-modules/events/docs/adr/0001-0005`.
- `CONTEXT-MAP.md` at repo root has been updated to include Events and Gamification bounded contexts with dependency diagram.
- `app-modules/events/CONTEXT.md` contains the full domain glossary and state machine reference.
- Future phases (post-MVP): magic_link/geolocation check-in, CFP/submissions, agenda display, networking, referral, payment integration, no-show penalties.

### LINKED ISSUE #245: feat(events): numeric code check-in
## Parent

#237

## What to build

Implement numeric code check-in: organizer generates a short-lived code bound to a specific date, participant enters it to check in. Also supports bot-triggered check-in via domain events.

**Admin panel — Code management:**
- "Generate Check-in Code" action on Event resource
- Fields: code (auto-generated or manual), event_date, valid_from, valid_until, max_uses (optional)
- List of active/expired codes per event with usage stats
- Code format: 4-6 digit numeric string

**Extend `CheckInAction` (numeric_code path):**
- Validates: code exists for this event + event_date
- Validates: `now() BETWEEN valid_from AND valid_until`
- Validates: `uses_count < max_uses` (if max_uses is not null)
- Atomically increments `uses_count` on the code record
- Creates check-in record with method=numeric_code, payload={code: "4829"}
- Same enrollment status transition logic as manual (confirmed → checked_in on first, record-only on subsequent)
- Dispatches `ParticipantCheckedIn`

**App panel — Code entry:**
- "Enter Check-in Code" form visible when enrollment is `confirmed` or `checked_in` (multi-day)
- Input field for the numeric code
- Validation errors: "Code expired", "Code already used maximum times", "Invalid code for today"
- Success confirmation on valid check-in

**Error handling:**
- Invalid code: 422 with "Invalid check-in code"
- Expired code (valid_until passed): 422 with "Code has expired"
- Exhausted code (uses_count >= max_uses): 422 with "Code has reached maximum uses"
- Wrong date (code is for a different event_date): 422 with "Code is not valid for today"

## Acceptance criteria

- [ ] Organizer can generate numeric codes bound to a specific event date
- [ ] Participant can enter code in App panel and get checked in
- [ ] Code validation checks: existence, date match, time window, max uses
- [ ] `uses_count` is incremented atomically on successful check-in
- [ ] Expired/exhausted codes are rejected with clear error messages
- [ ] Multi-day: participant can use different codes on different days
- [ ] Check-in record stores method=numeric_code with code in payload
- [ ] `ParticipantCheckedIn` domain event dispatched
- [ ] Feature tests: valid code, expired code, exhausted code, wrong date, concurrent usage
- [ ] Pint passes

## Blocked by

- #240 (RSVP Enrollment — needs confirmed enrollments)

### PR DESCRIPTION
**Pull request by davicbtoliveira**
Closes #245

### Summary

Implements self-service check-in via short-lived numeric codes announced by the organizer. Participants enter the code on the event detail page to check themselves in without organizer intervention.

### What was done

**Domain layer**
- Added `revoked_at` column to `events_check_in_codes` (soft revoke, per ADR-0003)
- Created `NumericCodeCheckInDTO` carrying enrollment, code, and eventDate
- Created `NumericCodeCheckInAction` with 5-step validation pipeline: existence → date binding → expiry/revoked → max uses → atomic increment + delegation to core `CheckInAction`
- Added 4 new `CheckInException` factory methods: `invalidCheckInCode`, `checkInCodeExpired`, `checkInCodeExhausted`, `checkInCodeWrongDate`
- Translated error messages in en/pt_BR

**Admin panel**
- Added `CheckInCodesRelationManager` on the Event edit page
- Generate codes with 4 or 6 digit length selector, auto-generated read-only code
- `event_date` defaults to the event's starts_at, constrained to the event date range
- Revoke action with confirmation (sets `revoked_at`)

**App panel**
- Created `NumericCodeCheckIn` Livewire component embedded in the event detail page
- Code input visible when enrollment is `confirmed` or `checked_in`
- Validation errors surface as inline messages below the input

**Tests**
- 8 feature tests: valid check-in, invalid code, wrong date, expired, revoked, max uses exhausted, atomic increment, outside event range

### Quality
- Pint: passed
- PHPStan: passed (0 errors)
- All events tests: 86/86 passed (0 regressions)

### ADR
- ADR-0006 documents all design decisions

### CODE REVIEW

File: app-modules/panel-app/src/Livewire/Events/NumericCodeCheckIn.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public function enrollment(): ?Enrollment
+    {
+        return Enrollment::query()
+            ->where('event_id', $this->eventId)
+            ->where('user_id', auth()->id())
+            ->first();
+    }
+
+    #[Computed]
+    public function canCheckIn(): bool
```
Comment (hefeus) [inline app-modules/panel-app/src/Livewire/Events/NumericCodeCheckIn.php:45]: O componente de checkin por codigo está mostrando mesmo em eventos de checkin QRCode. Precisa de uma validação para mostrar apenas em eventos de codigo numerico

**Review by YuriSouzaDev — APPROVED** [review]

**Comment by sirelves** [comment]
Fiz uma passada focada em seguranca, consistencia e comportamento em producao. A feature esta bem encaminhada, mas eu seguraria o merge por alguns pontos:

1. `starts_at` do codigo nao e validado no check-in

Arquivo: `app-modules/events/src/CheckIn/Actions/NumericCodeCheckInAction.php`, perto das validacoes de `expires_at`/`revoked_at`.

O admin salva `starts_at` como inicio da janela de validade, mas `NumericCodeCheckInAction` so valida `expires_at` e `revoked_at`. Na pratica, um codigo criado para comecar mais tarde ja pode ser usado antes do horario se o participante souber/adivinhar o valor.

Sugestao: validar tambem `starts_at`, por exemplo rejeitando quando `now()->lt($codeRecord->starts_at)`.

2. Falta protecao contra brute force no fluxo de codigo numerico

Arquivos:
- `app-modules/panel-app/src/Livewire/Events/NumericCodeCheckIn.php`, metodo `checkIn()`
- `app-modules/panel-admin/src/Filament/Resources/Events/RelationManagers/CheckInCodesRelationManager.php`, seletor de 4/6 digitos

O componente Livewire permite tentativas repetidas de `checkIn()` e o admin permite codigos de 4 digitos. Mesmo exigindo inscricao confirmada, 10k combinacoes e um espaco pequeno para tentativa automatizada.

Sugestao: adicionar rate limit por `user_id:event_id` (e possivelmente IP) antes de consultar o codigo, alem de validar formato/tamanho do codigo no componente.

3. Busca do codigo nao filtra por `event_date`

Arquivos:
- `app-modules/events/src/CheckIn/Actions/NumericCodeCheckInAction.php`, query inicial de `CheckInCode`
- `app-modules/events/database/migrations/2026_05_16_200006_create_events_check_in_codes_table.php`, constraint `unique(event_id, event_date, code)`

A tabela permite o mesmo `code` para datas diferentes do mesmo evento, mas a query busca apenas por `event_id + code` e usa `first()`. Em evento multi-dia, se o mesmo codigo existir em datas diferentes, pode pegar o registro errado e rejeitar um codigo valido para o dia atual.

Sugestao: incluir `event_date` na query inicial, alinhando com a constraint da tabela e com a regra de negocio.

4. O codigo exibido no preview do admin pode nao ser o codigo salvo

Arquivo: `app-modules/panel-admin/src/Filament/Resources/Events/RelationManagers/CheckInCodesRelationManager.php`, campo `code_preview` e callback `using()`.

`code_preview` gera um valor no formulario, mas esta com `dehydrated(false)`. Depois, no `using()`, outro codigo e gerado novamente. Isso pode fazer o organizador ver/anunciar um codigo diferente do que foi persistido. Alem disso, o preview sempre usa 6 digitos, mesmo quando o usuario seleciona 4.

Sugestao: gerar uma unica vez o codigo que sera salvo, ou atualizar o preview de forma reativa com o mesmo valor usado na persistencia.

Tambem vale adicionar testes cobrindo esses casos: codigo antes de `starts_at`, colisao/reuso de codigo em evento multi-dia, rate limit/tentativas repetidas e consistencia entre preview e valor persistido.

**Comment by sirelves** [comment]
Boa, os pontos anteriores parecem enderecados nesta ultima versao.

Fiquei com uma duvida sobre a mudanca em `config/database.php`:

```php
'timezone' => 'UTC',
```

Essa alteracao muda o timezone da conexao Postgres de forma global, fora do escopo direto do check-in numerico. Como o app usa `APP_TIMEZONE` com default `America/Sao_Paulo`, vale confirmar se essa mudanca e realmente necessaria para este PR e se nao pode afetar outras queries/serializacoes de `timestampTz` no restante da aplicacao.

Se a intencao era apenas estabilizar os testes/validacoes desse fluxo, talvez seja melhor resolver localmente no teste ou na comparacao de datas, para evitar impacto transversal.

**Comment by davicbtoliveira** [comment]
Corrigido, codex tava fritando ontem pra corrigir 2 testes q nao passavam e acabou trocando a timezone do banco 🤡 @sirelves

**Review by sirelves — APPROVED** [review]
LGTM. Obrigado pelos ajustes!

**Review by GabrielFVDev — APPROVED** [review]
brabo dms! 
LGTM

**Review by YuriSouzaDev — APPROVED** [review]

File: app-modules/events/tests/Feature/EventResourceTest.php [hunk 2]
```diff
@@ -141,6 +143,30 @@
...
+            'digits' => '4',
+            'code_preview' => '1234',
+            'event_date' => now()->toDateString(),
+            'starts_at' => now(),
+            'expires_at' => now()->addHours(2),
+            'max_uses' => '',
+        ])
+        ->assertHasNoTableActionErrors();
+
+    expect(CheckInCode::query()->where('event_id', $event->id)->sole()->code)->toBe('1234');
```
Comment (danielhe4rt) [inline app-modules/events/tests/Feature/EventResourceTest.php:167]: Aprendi negócio novo!!
